src/restruct_poc/constraints/calculate_constraint_residual_gradient.hpp

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include <Kokkos_Core.hpp>
+
+#include "calculate_cylindrical_constraint.hpp"
+#include "calculate_fixed_bc_constraint.hpp"
+#include "calculate_prescribed_bc_constraint.hpp"
+#include "calculate_rigid_constraint.hpp"
+#include "calculate_rotation_control_constraint.hpp"
+#include "constraints.hpp"
+
+#include "src/restruct_poc/math/quaternion_operations.hpp"
+#include "src/restruct_poc/math/vector_operations.hpp"
+#include "src/restruct_poc/types.hpp"
+
+namespace openturbine {
+
+struct CalculateConstraintResidualGradient {
+    Kokkos::View<ConstraintType*>::const_type type;
+    Kokkos::View<size_t* [2]>::const_type node_index;
+    Kokkos::View<size_t* [2]>::const_type row_range;
+    Kokkos::View<size_t* [2][2]>::const_type node_col_range;
+    Kokkos::View<double* [3]>::const_type X0_;
+    Kokkos::View<double* [3][3]>::const_type axis;
+    View_N::const_type control;
+    View_Nx7::const_type constraint_u;
+    View_Nx7::const_type node_u;
+    View_N Phi_;
+    Kokkos::View<double* [6][12]> gradient_terms;
+
+    KOKKOS_FUNCTION
+    void operator()(const int i_constraint) const {
+        if (type(i_constraint) == ConstraintType::kFixedBC) {
```

Comment:
Is there a reason this isn't a `switch` statement?

Reply:
Style.  It's not obvious which one will be faster (or that the speed difference will be significant).  I tend to prefer `if` statements, since they have no chance for fall-through behavior (thus are more readable and less bug prone) and compilers/hardware tend to be well optimized for them.

Reply:
Not sure I agree with the readability comment -- I find `switch` statements to be cleaner and more readable than an equivalent nested `if-else if-else` block. The cpp-coreguidelines appear to have the same outlook.

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-switch-if

Reply:
My comment on fall-through is about the need for a `break` at the end of each case.  For example
```
switch(type) {
  case type_1 :
    std::cout << 1;
  case type_2:
    std::cout << 2;
    break;
  case type_3:
    std::cout << 3;
    break;
}
```

If my `type` is `type_1`, then it will print 12, but if my `type` is `type_2` it will just print 2.  I know that you can write this type of code intentionally, but I would contend that it is harder to reason about - not only do I have to remember that the fall through happens, but I have to determine if it was intentional.  Compare this to:
```
if(type == type_1) {
  std::cout << 12;
}
if(type == type_2) {
  std::cout << 2;
}
if(type == type_3) {
  std::cout << 3;
}
```
Here, I know exactly what is happening for each case and there is no ambiguity of if the behavior for `type_1` is a bug.  

Looking at the page you linked, it does mention this in ES.78.  I agree with "always use `break` or `[[fallthrough]`. 
 However, I disagree with their suggestion of using multiple labels for a single statement.  Coming back later, it is hard to determine if you really meant for fall through or wanted to do nothing for that case and forgot a `break`.
 
 However, if both of you agree that we should you switch statements in this case, I will make the change.

Reply:
You can leave it as `if` statements. I'm more used to languages that don't use the implicit fallthrough in their switches like Fortran and Go. It is an easy construct to mess up in C/C++.

Reply:
Not a huge fan of the fall-through behavior myself either and tend to use this with caution and always with `break` everywhere else to make intentions clear. I personally think `switch` is neater than clunky-looking conditional branching in many instances.